tmt/utils.py

```diff
@@ -6560,7 +6561,10 @@ def default_template_environment() -> jinja2.Environment:
     Adds common filters, and enables block trimming and left strip.
     """
 
-    environment = jinja2.Environment()
+    # S701: `autoescape=False` is dangerous and can lead to XSS.
+    environment = jinja2.Environment(
+        autoescape=jinja2.select_autoescape(enabled_extensions='html', default_for_string=False)
```

Comment:
Here it's not being used for xml, no?  
Not sure about explicit `default=False` tbh. Do you think it's realistic they would change it in the future?

Reply:
> Here it's not being used for xml, no?

Probably not right now, but the escaping mechanism is similar enough for `html` and `xml` I think it's fine to include it rather than being surprised later on. It may be used for `junit` and other reporting mechanism later on, maybe even for plugins.

> Not sure about explicit `default=False` tbh. Do you think it's realistic they would change it in the future?

Jinja changing it like that, realistically not, it would be quite a dick move. But it's also a bit documentative that we really do not want to autoescape in general, particularly on how it impacts the general usecase of the template in the context of `fmf` and `yaml` files.

For the record if you want to reproduce how this affects, here is one yaml file that bit me when I've tried to add that:
```yaml
vars+:
  Extensions:
    - extras: " \" 1E-12\" "
```
Expanded in a Jinja template that basically used `{{ extras }}`. There the `\"` quote marks got changed to html escaped values.

Reply:
If it's necessary to not apply autoescaping just to some templates, then be aware that `render_template()` and `default_template_environment()` do not know what they will be rendering, and they may render various outputs: `report/html` uses them to render actual HTML content, and `export -h rest` to render a ReST representation of a story. Then there's the `export -h template` with a fully arbitrary template controlled by a user, it may be a plain-text file or XML, we don't know. Plus, often the environment is not used for loading the template itself, i.e. no chance for it to see the file extension (`return render_template(template_filepath.read_text(), ...). Enabling it here probably won't have the effect you might expect, and may have side effects.

Reply:
I would maybe turn it off, note it somewhere and make sure users would be aware of it... Or is there a way to make it enabled & correct for all possible templates? We already use it to render asciidoc and rest as well as HTML and neither of those two functions is aware of the nature of those templates...

Reply:
With the `enabled_extensions=['html', 'xml']` it should only use it on `.html`/`.xml` output files. Definitely needs to be tested though. I do see value in having it enabled for things like `report/html`

Reply:
Sure, I just point out that the environment knows nothing about the context. See https://github.com/teemtee/tmt/blob/main/tmt/steps/report/html.py#L100, this is how the template is rendered and saved. Environment is not aware of the destination file or its suffix.

I don’t dispute the value of autoescaping, just the way it would be enabled - there are no files the environment could inspect and decide based on their suffixes. `report/html` needs to tell to `render_template()` & co. “Guys, we’re rendering HTML, enable autoescaping, chop chop”. I suppose it would become a new parameter for the rendering family, would that make sense?

Reply:
> autoescape
>
>    If set to True the XML/HTML autoescaping feature is enabled by default. For more details about autoescaping see Markup. As of Jinja 2.4 this can also be a callable that is passed the template name and has to return True or False depending on autoescape should be enabled by default.

My reading of this is that:
- `template_name` is a string -> use `default_for_string`
- `template_name` has extension -> check `enabled/disabled_extensions` -> fallback to `default`

Reply:
@happz Not sure about keeping it off, based on what's being written in the Jinja API doc: "However autoescaping is not yet enabled by default though this will most likely change in the future. It’s recommended to configure a sensible default for autoescaping."
wdyt?

@LecrisUT I remembered why I only put html there - if we add 'xml' (or leave default, which is `['html', 'htm', 'xml']`), generating the docs fails.

Reply:
> @LecrisUT I remembered why I only put html there - if we add 'xml' (or leave default, which is `['html', 'htm', 'xml']`), generating the docs fails.

Actually that is great news. we can see exactly how this affects and if it's `template_name` vs `output_file`, etc. The `xml` part is unexpected for the site generation and would be interesting to hunt down what is going there.

Reply:
fwiw, it fails on url link: `tmt/docs/spec/hardware.rst:167: ERROR: Unknown target name: "pci id repository &lt;https://pci-ids.ucw.cz/&gt;".`

Reply:
> it seems to me you see some connection
> Can you explain that to me like I'm five?

@happz Don't know about others, but I thought it was implied clearly that I have no idea what's going on. 
I'm going through the templates code now and slowly starting to understand, but still - you telling me how to deal with the flake8 error is all I can hope for. Just trying to help.

The jinja doc imho makes it clear not using autoescape is bad (as well as the flake8-bandit), but I do not know how to deal with things like "from url", so if you say autoescape=False is the best way to go, at least for now all things considered, I'm happy :)

Reply:
Actually I think there is an issue with the implementation, but I need to look more closely. A few notes:
- `loader=FileSystemLoader()` should be used
  Jinja supports splitting a template into multiple files, and this loader is needed to be able to read those templates
- `env.get_tempalte(Path())`
  This should be coupled with the usage of `loader` above
- Looking with one implementation I have [here](https://github.com/LecrisUT/fmf-jinja/blob/66d243830871f8a5eecab718ae53f69e23f1d68c/src/fmf_jinja/template.py#L91-L96) (`autoescape` is wrong over there), indeed nowhere does it have access to the output file, only the template
- Regarding `.html.j2` vs `.j2.html`, if upstream doesn't support it, we should be able to write the callable to support that from their implementation.

> fwiw, it fails on url link: `tmt/docs/spec/hardware.rst:167: ERROR: Unknown target name: "pci id repository &lt;https://pci-ids.ucw.cz/&gt;".`

Is that with `default_for_string=False`? That is when running `sphinx` or which command? Should be able to traceback that and see in the debugger what produced that.

Reply:
> I for one would leave such a task for a standalone patch. It won't be an oneliner and it would go well beyond a mere addition of ruff rules. Worth fixing though, unless you guys would work it, I would, to get the escaping right eventually.

Agreed let's revert and comment out (`S701`?) as a planned improvement here. I've opened an issue to track it: https://github.com/teemtee/tmt/issues/2873

Reply:
@happz Feel free to rewrite the comment if needed :)